aya/src/programs/fentry.rs

```diff
@@ -0,0 +1,78 @@
+//! fentry programs.
+use thiserror::Error;
+
+use crate::{
+    generated::{bpf_attach_type::BPF_TRACE_FENTRY, bpf_prog_type::BPF_PROG_TYPE_TRACING},
+    obj::btf::{Btf, BtfError, BtfKind},
+    programs::{load_program, utils::attach_btf_id, LinkRef, ProgramData, ProgramError},
+};
+
+#[derive(Debug)]
+#[doc(alias = "BPF_TRACE_FENTRY")]
+#[doc(alias = "BPF_PROG_TYPE_TRACING")]
+pub struct FEntry {
+    pub(crate) data: ProgramData,
+}
+
+#[derive(Debug, Error)]
+pub enum FEntryLoadError {
+    #[error(transparent)]
+    Btf(#[from] BtfError),
+
+    #[error(transparent)]
+    Program(#[from] ProgramError),
+}
+
+/// Marks a function as a fentry eBPF program that can be attached to almost
+/// any function inside the kernel. The difference between fexit and kretprobe
+/// is that fexit has practically zero overhead to call before kernel function.
+/// fentry programs can be also attached to other eBPF programs.
+///
+/// # Minimumm kernel version
+///
+/// The minimum kernel version required to use this feature is 5.5.
+///
+/// # Examples
+///
+/// ```no_run
+/// # #[derive(thiserror::Error, Debug)]
+/// # enum Error {
+/// #     #[error(transparent)]
+/// #     BtfError(#[from] aya::BtfError),
+/// #     #[error(transparent)]
+/// #     FEntry(#[from] aya::programs::FEntryLoadError),
+/// #     #[error(transparent)]
+/// #     Program(#[from] aya::programs::ProgramError),
+/// #     #[error(transparent)]
+/// #     Bpf(#[from] aya::BpfError),
+/// # }
+/// # let mut bpf = Bpf::load_file("ebpf_programs.o")?;
+/// use aya::{Bpf, programs::FEntry, BtfError, Btf};
+/// use std::convert::TryInto;
+///
+/// let btf = Btf::from_sys_fs()?;
+/// let program: &mut FEntry = bpf.program_mut("filename_lookup").unwrap().try_into()?;
+/// program.load("filename_lookup", &btf)?;
+/// program.attach()?;
+/// # Ok::<(), Error>(())
+/// ```
+impl FEntry {
+    /// Loads the program inside the kernel.
+    ///
+    /// See also [`Program::load`](crate::program::Program::load).
+    ///
+    /// # Arguments
+    ///
+    /// * `fn_name` - functionn name inside the kernel to attach
+    /// * `btf` - btf information for the target system
+    pub fn load(&mut self, fn_name: &str, btf: &Btf) -> Result<(), FEntryLoadError> {
```

Comment:
This should return `ProgramError`, which should have a Btf variant if it doesn't have one already. (I think this change needs to be done in Lsm too - obvs not necessarily as part of this PR).

We don't want to expose too many error types as public APIs, the more public types, the harder to maintain and change in the long term.

Reply:
`ProgramError` doesn't have a straightforward `#[from] BtfError` variant, but we rather have a custom `BtfTracePointError` wrapping it which is defined here:

https://github.com/aya-rs/aya/blob/f56dd0a70b36a97036eb9447efa20f0e1c93c8d7/aya/src/programs/tp_btf.rs#L57-L63

and then included in `ProgramError`

https://github.com/aya-rs/aya/blob/f56dd0a70b36a97036eb9447efa20f0e1c93c8d7/aya/src/programs/mod.rs#L173-L175

That doesn't really make sense to me and I would rather just wrap `BtfError` directly in `ProgramError`, which is what I did in the last push. :) Let me know if you agree with that.